source/service.go

```diff
@@ -307,7 +307,22 @@ func (sc *serviceSource) extractHeadlessEndpoints(svc *v1.Service, hostname stri
 
 			for _, headlessDomain := range headlessDomains {
 				var ep string
-				if sc.publishHostIP {
+				if sc.publishHostIP && externalIP {
+					node, err := sc.nodeInformer.Lister().Get(pod.Spec.NodeName)
+					if err != nil {
+						return nil
+					}
```

Comment:
Consider checking [`errors.IsNotFound`](https://sourcegraph.com/github.com/kubernetes/apimachinery/-/blob/pkg/api/errors/errors.go#L478-L482) here. If the node is now absent, should we skip further processing of this pod? Are we checking for such absence higher up the call chain?

Reply:
I don't think we're checking for absence higher up the call chain.

Reply:
It's possible that we could see that error occur in two different scenarios:
- The node existed, something placed the pod onto it, but then the node was deleted.
- The node is new, and our queue worker for pods is responding to a watch firing before our cache has picked up the node's arrival.

Reply:
I've updated the code some. I'm not sure if I've addressed your concern or not. What would your expectation be for this code?

Reply:
I'm assuming we can stop processing, because it will get picked up on the next loop. Perhaps I'm wrong?

Reply:
By my reading of the code thus far, we won't come back through to respond to the node's presence changing, because we only maintain the cache and read it on demand. There are no event handlers bound [that do anything significant](https://github.com/kubernetes-sigs/external-dns/blob/b093524c0b8c9db30284fa2363a5ef40a1d3fc73/source/node.go#L68-L75).

I suggest distinguishing this "not found" error from others, logging a distinguished message for that case, and skipping the pod, carrying on with any others nominated by the _Endpoints_ entries. As it stands, the containing function doesn't allow any indication to its callers that it couldn't finish its work completely, where the pod should be queued again for later consideration.